### ISSUE TITLE
Customizable Album image ordering

### ISSUE THREAD

**Issue by arahayrabedian** [agile-leaf/50mm#10]
I understand this is introducing complexity, but hear me out

Basically when I go on a trip I have two cameras on me, my DSLR and my phone, the filenames on these (and their modification dates/times) don't necessarily do my album a good service.

Would you be open to the idea of having an `ordering.txt` file dumped in the bucket in order to impose an ordering? does this violate the philosophy?

I'm thinking something like:
```
DSC_2222.JPG
DSC_3333.JPG
DSC_1111.JPG
DSC_4444.JPG
DSC_8888.JPG
...
```
of course this also brings many edge cases like how to handle missing keys, or extra keys, etc., I'd say that's up to the maintainer/implementer to decide on.

**Comment by arahayrabedian** [comment]
stupid me realises that perhaps even an ugly list of keys in the config files of an album's section may work

**Comment by theonejb** [comment]
I had the same idea when I was building the thing, figuring out how to decide on an order without a DB in the backend. However at the time I just wanted to launch without spending time on additional features, so I instead settled for using my album manager (Lightroom) just rename the files in the order I wanted.

But this idea still appeals to me as a relatively good one. It doesn't need a DB, so 50mm is kept stateless, and it's possible to create that list with just a texteditor.

However, the edge cases you mentioned; plus a bunch of others we haven't thought of yet, will need to be managed. And we'll probably need a way to allow users to create the ordering with an interface of some sort, since manually doing it by hand is difficult. In-fact, it might be easier to just rename the image files on the system before uploading if the `ordering.txt` is to be created by hand, since at least then you can see a preview of the images.

Just a few thoughts. Let's hash this out and see if it's something we can build. Looking forward to hearing your thoughts on this.

**Comment by theonejb** [comment]
Although I would be hesitant in adding the ordering to the config files, since that violates on of the philosophy of 50mm; it should be stateless once an album is configured, and all data for the images (and thus the ordering I guess) should come from the S3 bucket.

**Comment by arahayrabedian** [comment]
the problem is I've already uploaded my photos, my personal opinion is that it's not nice to force people to mold their file naming conventions to a product, for some people the S3 bucket may also be acting as a backup, and recovering the names would be inconvenient.

That said, requiring somebody to write up an ordering file is indeed annoying, my personal plan is to order by exif created-at and dump the list, I'd people capable of setting up 50mm would be capable doing the same (to start with)

I'd expect later JS to be able to edit the file, but that's beyond my personal capabilities.

As for ordering, my proposals (not entirely thought through) would be, in preferred order:
1) go for a full on yaml (or similar) config, capable of defining the cover photo and thumbnails to select (this just came to me as i was writing, it was second preferred and is now top choice)
1) use the sorted list for the first files, then fallback to bucket ordering for everything else
1) ignore the file if it does not match the list of files exactly and fall back to bucket ordering (break easily and hard)

I'm not sure how these play with the philosophy, but IMO it keeps it as "point it at a bucket and it'll run, but you can add in some stuff to customize if you really need to"

what do you think?

**Comment by theonejb** [comment]
I like proposal 1, with a YAML or similar config. I was thinking, what if we just stored a config.yml in the S3 bucket? Instead of having to restart the server on every config change, why not just store it in the bucket.

Then load the config file on first bucket access and cache for some time; similar to what we are doing with the photos list.

If the config fails to load; we just assume it's not there.

This has the benefit of allowing us to develop a separate JS based frontend to edit the config and write it back to the bucket, without needing access to the server configs.

However, I'm not sure what format we can use for the ordering. Here are some thoughts:

 - Requiring all photos to be in the ordering list means that every time I upload a new image, my ordering breaks until I add it to the list. This breaks the "Upload via FTP and everything works" design
 - Allowing partial ordering, by only listing the images you want explicitly ordered. In this case, where do the rest of the images in the bucket go?
 - Order by EXIF created by date by reading the list on album access, but that would be horribly slow.

And in all cases where the user can configure the ordering by a list of photo keys, how do you handle photo names that are not in the album? Error out or skip those entirely.

**Comment by arahayrabedian** [comment]
yeah i meant tossing the yaml in the bucket as well :)

>Requiring all photos to be in the ordering list means that every time I upload a new image, my ordering breaks until I add it to the list. This breaks the "Upload via FTP and everything works" design

fair point, we can drop this from consideration

>Allowing partial ordering, by only listing the images you want explicitly ordered. In this case, where do the rest of the images in the bucket go?

my intention was to take files listed, display those first, and then whatever's left from the bucket index read we can put in the traditional ordering, for example, if you have in your ordering configuration (skipping yaml formatting as we don't know that looks like yet):
```
6.jpg
5.jpg
8.jpg
4.jpg
```
and in your actual bucket you have files from 1-100 in numeric order, we could then do:
```
6.jpg
5.jpg
8.jpg
4.jpg
1.jpg
2.jpg
3.jpg
7.jpg
9.jpg
10.jpg
...
100.jpg
```

and if an image doesn't exist but is defined in the file, we can silently drop it or leave a placeholder, your call.

the only issue with this is, if you've set up your album ordering and you add in a new photo without updating your config, it'll end up being appended to the end until you update the config, which may be a little cumbersome until things settle.

**Comment by theonejb** [comment]
I like the partial ordering idea, given it's the closest to the kind of use case I'd have with manual ordering; basically display these few photos first because I think they are the most important ones.

As for having new photos appear at the end, they would do so with or without custom ordering in place, so while not perfect it's not a problem that would be introduced by this feature.

I guess we can go ahead and start implementing this. Do you want to take this on or should I?

**Comment by arahayrabedian** [comment]
I wouldn't mind taking it on but in a few days, just to clear up a minor thing:

I was actually thinking of the yaml having a few options, so in full it looks more like:
```yaml
cover: 77.jpg
thumbnails: # can't remember how many you need
  - 55.jpg
  - 44.jpg
  - 45.jpg
  - 66.jpg
  - 88.jpg
  - 4.jpg
ordering:
  - 1.jpg
  - 2.jpg
  - 4.jpg
  - 3.jpg
  - 5.jpg
  - 6.jpg
...
```
In my use case I don't usually want the first photo to be the cover photo (though I prefer the album itself to be in date taken order)

basically my intention is to kill multiple birds with one stone, if this is too much and you'd prefer only an ordering system, I'd still be happy with that.

edit: each section should be optional and having flexible fallbacks as previously discussed

**Comment by theonejb** [comment]
This is perfect.

**Comment by arahayrabedian** [comment]
awesome, let me finish the other PR and i'll get on to this one.

**Comment by arahayrabedian** [comment]
i'mma take a stab at this

**Comment by theonejb** [comment]
Go for it! Can't wait to see what you come up with.

Also, sorry for _still_ not having done that code review. Will get to it in a bit, just so busy right now!

**Comment by theonejb** [comment]
Closed by pull request #14
